[WebAudio] Ensure AudioBufferSourceNode doesn't read past the input source when not looping

aperezdc opened the pull request:
#### c1ad8e16b414fdeb50f5220cd7cd574f179b6096
<pre>
[WebAudio] Ensure AudioBufferSourceNode doesn&apos;t read past the input source when not looping
<a href="https://bugs.webkit.org/show_bug.cgi?id=287477">https://bugs.webkit.org/show_bug.cgi?id=287477</a>

Reviewed by Darin Adler and Chris Dumez.

Add an early exit from AudioBufferSourceNode::renderFromBuffer()
instead of reading past the source buffer when m_isLooping is false.

The read cursor in AudioBufferSourceNode may have been set e.g. due to a
previous ::renderFromBuffer() call to a position past the end of the
source buffer. If the node has m_isLooping set the existing checks will
reset the cursor to the beginning of the loop; but if not looping then
we may attempt to read past the end of the source buffer.

* Source/WebCore/Modules/webaudio/AudioBufferSourceNode.cpp:
(WebCore::AudioBufferSourceNode::renderFromBuffer):

Canonical link: <a href="https://commits.webkit.org/290282@main">https://commits.webkit.org/290282@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/81af2e31bd4e4f57b78516e67bb07e36d8d4d009

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89560 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/131/builds/9087 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/55/builds/44419 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/94552 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/40327 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/91612 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/130/builds/9476 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/123/builds/17366 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/94552 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/40327 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92561 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/130/builds/9476 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/55/builds/44419 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/94552 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/130/builds/9476 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/55/builds/44419 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/39433 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/130/builds/9476 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/55/builds/44419 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/96380 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/16742 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/123/builds/17366 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/96380 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/16997 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/55/builds/44419 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/96380 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/55/builds/44419 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/9909 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14041 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/16755 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/22069 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/16496 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/19947 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/18278 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro reviewed (commented):
I'm a little surprised that you're comparing doubles here instead of integers. The array indexes are ultimately cast to integers. Is it not possible to do this check further down, once you have integers to work with, so that we don't have to think about floating point rounding? If this is the best place to do the check, how do we know it's safe from potential rounding issues?

darinadler commented on Source/WebCore/Modules/webaudio/AudioBufferSourceNode.cpp:
```diff
@@ -249,7 +249,7 @@ bool AudioBufferSourceNode::renderFromBuffer(AudioBus* bus, unsigned destination
 
     // Sanity check that our playback rate isn't larger than the loop size.
-    if (std::abs(pitchRate) > virtualDeltaFrames)
+    if (std::abs(pitchRate) > virtualDeltaFrames || m_virtualReadIndex >= virtualMaxFrame)
```
Comment no longer matches the code below it. Could use a separate `if` and leave the comment as is, or change the comment, or delete the comment since it says the same thing the code does.

aperezdc replied:
Given that the same check is done in the `if` right before this one, maybe it makes more sense to rewrite the previous one, that currently is:

```cpp
    if (m_isLooping && m_virtualReadIndex >= virtualMaxFrame) {
        // Adjust cursor to loop back.
    }
```

into something like this instead:

```cpp
    if (m_virtualReadIndex >= virtualMaxFrame) {
        // Early exit to avoid going past the end of the source buffer.
        if (!m_isLooping)
            return false;

        // Adjust cursor to loop back.
     }
```

What do we think about this option?

jernoble replied:
LGTM

aperezdc replied:
Done, changed the patch in this way 👌🏼

darinadler commented on Source/WebCore/Modules/webaudio/AudioBufferSourceNode.cpp:
```diff
@@ -241,9 +241,13 @@ bool AudioBufferSourceNode::renderFromBuffer(AudioBus* bus, unsigned destination
+            return false;
+
+        // If we're looping and the offset (virtualReadIndex) is past the end of the loop, wrap back to the
+        // beginning of the loop. For other cases, nothing needs to be done.
```
Can we remove "For other cases, nothing needs to be done." here? With the comment inside the `if` statement, it no longer makes much sense nor does it seem like it needs to be said. I think the only part of the comment that is non-obvious, so worth writing, is "wrap back to the beginning of the loop". The rest is just a comment that says exactly what the code says ("if we’re looping" is the same as checking `m_isLooping`, past the end of the loop is the same thing as `m_virtualReadIndex >= virtualMaxFrame`), so not my favorite, although our web audio code has a lot of those kinds of comments. And note that before this code was documenting what was under it, now it’s referring to what’s above it, so less clear.

aperezdc replied:
Of course, I'll amend before landing, thanks! 🙏🏼

aperezdc committed:
[WebAudio] Ensure AudioBufferSourceNode doesn't read past the input source when not looping
https://bugs.webkit.org/show_bug.cgi?id=287477

Reviewed by Darin Adler and Chris Dumez.

Add an early exit from AudioBufferSourceNode::renderFromBuffer()
instead of reading past the source buffer when m_isLooping is false.

The read cursor in AudioBufferSourceNode may have been set e.g. due to a
previous ::renderFromBuffer() call to a position past the end of the
source buffer. If the node has m_isLooping set the existing checks will
reset the cursor to the beginning of the loop; but if not looping then
we may attempt to read past the end of the source buffer.

* Source/WebCore/Modules/webaudio/AudioBufferSourceNode.cpp:
(WebCore::AudioBufferSourceNode::renderFromBuffer):

Canonical link: https://commits.webkit.org/290282@main
